File: line_nihe.cpp
#include <opencv2/opencv.hpp>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>

#define MAX_POINTS 10000

using namespace cv;

// 初始化摄像头
VideoCapture init_camera_or_video(int camera_id, const std::string& video_path) {
    VideoCapture cap(camera_id);
    if (!cap.isOpened()) {
        fprintf(stderr, "警告：摄像头 %d 打开失败，尝试打开视频文件：%s\n", camera_id, video_path.c_str());

        cap.open(video_path);
        if (!cap.isOpened()) {
            fprintf(stderr, "错误：无法打开视频文件 %s\n", video_path.c_str());
            exit(1);
        } else {
            printf("提示：已切换为本地视频输入\n");
        }
    } else {
        printf("提示：成功打开摄像头 %d\n", camera_id);
    }
    return cap;
}

// 提取边缘中点（每行一个平均 x）
int get_edge_points(Mat edge_img, int y_offset, int x_offset, Point* out_points) {
    int count = 0;
    for (int y = 0; y < edge_img.rows; y++) {
        uchar* row = edge_img.ptr<uchar>(y);
        int sum_x = 0, cnt = 0;
        for (int x = 0; x < edge_img.cols; x++) {
            if (row[x] > 0) {
                sum_x += x;
                cnt++;
            }
        }
        if (cnt > 0) {
            int avg_x = sum_x / cnt + x_offset;
            int real_y = y + y_offset;
            out_points[count++] = Point(avg_x, real_y);
            if (count >= MAX_POINTS) break;
        }
    }
    return count;
}

// RANSAC 拟合直线
int ransac_fit(Point* points, int num_points, float* out_k, float* out_b) {
    if (num_points < 2) return 0;  // 防止传入点过少

    int best_inliers = 0;
    float best_k = 0, best_b = 0;

    srand((unsigned int)time(NULL));

    for (int i = 0; i < 100; i++) {
        int idx1 = rand() % num_points;
        int idx2 = rand() % num_points;
        if (idx1 == idx2) continue;

        Point p1 = points[idx1];
        Point p2 = points[idx2];

        if (p1.x == p2.x) continue;

        float k = (float)(p2.y - p1.y) / (p2.x - p1.x);
        if (fabs(k) > 1.0f) continue; // ✅ 斜率过滤

        float b = p1.y - k * p1.x;

        int inliers = 0;
        for (int j = 0; j < num_points; j++) {
            Point p = points[j];
            float dist = fabs(k * p.x - p.y + b) / sqrt(k * k + 1);
            if (dist < 3.0)
                inliers++;
        }

        if (inliers > best_inliers) {
            best_inliers = inliers;
            best_k = k;
            best_b = b;
        }
    }

    *out_k = best_k;
    *out_b = best_b;
    return best_inliers;
}


int main() {
VideoCapture cap;
    //VideoCapture cap = init_camera_or_video(2,"/home/orangepi/rknn-cpp-Multithreading-main/output.mp4"); // 使用 2 号摄像头
    if (!cap.open("/home/orangepi/rknn-cpp-Multithreading-main/output.mp4")) {
        return -1;
    }
    while (true) {
        Mat img;
        cap >> img;
        if (img.empty()) {
            fprintf(stderr, "Error: Frame capture failed\n");
            break;
        }

        int h = img.rows;
        int w = img.cols;

        Mat gray, blur, edges;
        cvtColor(img, gray, COLOR_BGR2GRAY);
        GaussianBlur(gray, blur, Size(7, 7), 0);

        int roi_top = h / 2 - 120;
        int roi_bot = h / 2 + 120;
        if (roi_top < 0 || roi_bot > h) continue; // 检查边界
        Mat roi = blur(Range(roi_top, roi_bot), Range::all());
        
        imshow("roi", roi);
        
        Canny(roi, edges, 50, 150);

        Point all_points[MAX_POINTS];
        int total = 0;

        int w1 = w / 3;
        int w2 = 2 * w / 3;

        Mat left = edges(Range::all(), Range(0, w1));
        total += get_edge_points(left, roi_top, 0, all_points + total);

        Mat mid = edges(Range::all(), Range(w1, w2));
        total += get_edge_points(mid, roi_top, w1, all_points + total);

        Mat right = edges(Range::all(), Range(w2, w));
        total += get_edge_points(right, roi_top, w2, all_points + total);
        
        
        if (total < 2) {
            printf("警告：未检测到足够的边缘点用于拟合，跳过此帧\n");
            imshow("Line Fitting", img);
            if (waitKey(1) == 27) break;
            continue;
        }


        float k, b;
        ransac_fit(all_points, total, &k, &b);

        float y_left = k * w1 + b;
        float y_right = k * w2 + b;
        float gap = fabs(y_right - y_left);

        printf("斜率 k = %.4f, 中间间距 = %.2f 像素\n", k, gap);

        Mat out_img = img.clone();
        Point pt1(0, (int)(b));
        Point pt2(w, (int)(k * w + b));
        line(out_img, pt1, pt2, Scalar(0, 255, 0), 2);
        //打印所有点
        // 可视化所有拟合点（红色小圆点）
        for (int i = 0; i < total; i++) {
            circle(out_img, all_points[i], 2, Scalar(0, 0, 255), -1); // 红色
        }
        imshow("Line Fitting", out_img);
        if (waitKey(1) == 27) break; // 按 ESC 退出
    }

    cap.release();
    destroyAllWindows();
    return 0;
}

File: line_nihe.py
import cv2
import numpy as np

# 读取图像
img = cv2.imread('7777.png')
h, w = img.shape[:2]

# 转灰度 + 去噪
gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
blur = cv2.GaussianBlur(gray, (7, 7), 0)

# 只关注图像中间区域（垂直方向）
roi_top = h // 2 - 120
roi_bot = h // 2 + 530
roi_left=150
roi_right=1200
roi = blur[roi_top:roi_bot, :]
cv2.imshow('roi', roi)

# Canny边缘检测
edges = cv2.Canny(roi, 50, 150)

# 显示中间 ROI 边缘图像
cv2.imshow("Edges", edges)

# 原图副本，用于画点
out_img = img.copy()
# cv2.line(out_img, (0, roi_top), (w , roi_bot), (255, 0, 0), 12)
# print("roi_top",roi_top,"roi_bot",roi_bot)
line_points = []

# 分割区域
left_roi = edges[:, :w // 3]
middle_roi = edges[:, w // 3 : 2 * w // 3]
offset_middle =  w // 3  # 中间区域偏移回整图
right_roi = edges[:, -w // 3:]
offset_right = 2 * w // 3 # 右区域偏移回整图

cv2.imshow('left_roi', left_roi)
cv2.imshow('middle_roi', middle_roi)
cv2.imshow('right_roi', right_roi)

# 提取左、右边缘点
def get_edge_points(sub_edges, x_offset=0):
    points = []
    for y in range(sub_edges.shape[0]):
        row = sub_edges[y]
        edge_x = np.where(row > 0)[0]
        if len(edge_x) > 0:
            x_mean = int(np.mean(edge_x)) + x_offset
            y_real = y + roi_top
            points.append((x_mean, y_real))
    return points

left_points = get_edge_points(left_roi)
middle_points= get_edge_points(middle_roi,offset_middle)
right_points = get_edge_points(right_roi, offset_right)

# # 合并左右点
fit_points = left_points + middle_points + right_points



import random

def ransac_line_fit(points, num_iters=100, distance_threshold=3):
    best_inliers = []
    best_model = None

    for _ in range(num_iters):
        # 随机选2个点
        sample = random.sample(points, 2)
        (x1, y1), (x2, y2) = sample

        # 跳过垂直线避免除零
        if x2 == x1:
            continue

        # 拟合直线参数 y = kx + b
        k = (y2 - y1) / (x2 - x1)
        b = y1 - k * x1

        inliers = []
        for (x, y) in points:
            # 计算点到直线的距离
            distance = abs(k * x - y + b) / np.sqrt(k**2 + 1)
            if distance < distance_threshold:
                inliers.append((x, y))

        if len(inliers) > len(best_inliers):
            best_inliers = inliers
            best_model = (k, b)

    return best_model, best_inliers


# 可视化所有拟合点
for (x, y) in fit_points:
    cv2.circle(out_img, (x, y), 2, (0, 0, 255), -1)

# 用 RANSAC 拟合直线
(k, b), inliers = ransac_line_fit(fit_points)

# 拟合线端点
x_start = 0
y_start = int(k * x_start + b)

x_end = w
y_end = int(k * x_end + b)

# 在图像上画线
cv2.line(out_img, (x_start, y_start), (x_end, y_end), (0, 255, 0), 2)

# 输出斜率
print(f"斜率 k = {k:.4f}")

# 计算中间区域左右边界的 x 值
middle_left_x = w // 3
middle_right_x = 2 * w // 3

# 对应直线上的 y 值
y_left = k * middle_left_x + b
y_right = k * middle_right_x + b

# 中间的“高度间距”
middle_gap = abs(y_right - y_left)

print(f"中间区域直线 y 值差（间距）: {middle_gap:.2f} 像素")


# 可视化 inliers 点（拟合点）
for (x, y) in inliers:
    cv2.circle(out_img, (x, y), 2, (255, 0, 0), -1)

# 再次显示图像
cv2.imshow("Fitted Line", out_img)
cv2.waitKey(0)
cv2.destroyAllWindows()